src/StdString.cpp

```diff
@@ -0,0 +1,118 @@
+// I've reorganized StdString.h to be optimized for the ITGmania engine.
+//
+// A huge amount of code bloat was traditionally due to these functions being
+// inlined. Performance issues with RStrings are largely mitigated by migrating
```

Comment:
I don't understand this. Inline functions are typically faster because they don't incur the function calls which are a slow instruction. The decrease in binary size makes sense considering they were formerly inlined. That being said, I don't understand why this change would have such an impact on fps when all it seems like you're doing is moving to function implementations to a cpp file and then removing inline. Is the fps change more related to noexcept? (which I think could make more sense especially compared to removing inline). I'd like a more objective explanation as to the reasoning and not just citing "fps gain"

Reply:
> I don't understand this. Inline functions are typically faster because they don't incur the function calls which are a slow instruction. The decrease in binary size makes sense considering they were formerly inlined. That being said, I don't understand why this change would have such an impact on fps when all it seems like you're doing is moving to function implementations to a cpp file and then removing inline. Is the fps change more related to noexcept? (which I think could make more sense especially compared to removing inline). I'd like a more objective explanation as to the reasoning and not just citing "fps gain"

I'd be more than happy to talk about this! ☺️ 

Inline functions are not inherently fast, they can be faster but they can be slower too. In this way it's similar to how a #define macro works - the compiler expands that code where that function was called. That doesn't guarantee it would be faster though. If function overhead or function return overhead is significant, then it's highly beneficial to inline a function. 

The problem is that anything related to RString is used many, many times in the code base. That means anytime RString is being used, the code is being inlined. This is very bad because it's called VERY frequently: The size of the functions moved to StdString.cpp make up about 1k altogether. ~300kb of binary size is saved by this change (MSVC), and that includes whatever optimizations the compiler has made. 

The reason the inlining of these functions isn't helpful is because we know the engine is not CPU bound. I/O activity is extremely heavy regardless of CPU power (one of the reasons high end gaming setups do not provide a better ITGmania experience than some typical office PC). Lua is also being interpreted line by line which adds to I/O congestion. Additional evidence is that CPU usage is barely anything even on older CPU's. Inlining isn't helping us with RString because we are not having CPU issues or bottlenecks caused by the overhead of function calls. Inlining provides no help in an I/O bound situation. In fact it actually makes things worse because page faults are increased as the game must find its specific instance of whatever function, leading to memory thrashing. 

Of course inlining functions can prevent thrashing too, when used correctly. An example of this would be a function called in a tight loop. It's likely that the many function calls in GameLoop's main loops could benefit from inlining for example.

See my RageSoundMixBuffer PR, for example. You'll notice I moved both read functions as well as SetWriteOffset to be inline functions. These functions benefit from being inlined because they are very simple, and are more likely to be constrained by function calling overhead or pushing/popping the variables involved (more instructions are used on moving the variables than the function itself). 

Of course, noexcept also helps here too, because it allows the compiler to optimize differently than it would if it assumed  the functions might need to be able to throw errors. The bigger benefit is the significant reduction in disk I/O especially in-game. Think about all the RStrings in Actor and GameLoop alone, tons that run on every frame. That's a huge reduction in disk I/O which directly leads to better performance since less time is wasted reading from the executable.

Reply:
memory thrashing makes a lot of sense! Thanks for the explanation!